util/chunk/iterator.go

```diff
@@ -284,3 +303,72 @@ func (it *iterator4RowPtr) ReachEnd() {
 func (it *iterator4RowPtr) Len() int {
 	return len(it.ptrs)
 }
+
+// Error returns none-nil error if anything wrong happens during the iteration.
+func (it *iterator4RowPtr) Error() error {
+	return nil
+}
+
+// NewIterator4RowContainer create a new iterator for RowContainer
+func NewIterator4RowContainer(c *RowContainer) *iterator4RowContainer {
+	return &iterator4RowContainer{c: c}
+}
+
+type iterator4RowContainer struct {
+	c      *RowContainer
+	chkIdx int
+	rowIdx int
+	err    error
+}
+
+func (it *iterator4RowContainer) Len() int {
+	return it.c.NumRow()
+}
+
+func (it *iterator4RowContainer) setNextPtr() {
+	it.rowIdx++
+	if it.rowIdx == it.c.NumRowsOfChunk(it.chkIdx) {
+		it.rowIdx = 0
+		it.chkIdx++
+	}
+}
+
+func (it *iterator4RowContainer) Begin() Row {
+	it.chkIdx, it.rowIdx = 0, -1
+	return it.Next()
+}
+
+func (it *iterator4RowContainer) Next() Row {
+	if it.chkIdx >= it.c.NumChunks() {
+		it.ReachEnd()
+		return it.End()
+	}
+	it.setNextPtr()
+	return it.Current()
+}
+
+func (it *iterator4RowContainer) Current() Row {
+	if it.rowIdx < 0 || it.chkIdx >= it.c.NumChunks() {
+		return it.End()
+	}
+	row, err := it.c.GetRow(RowPtr{uint32(it.chkIdx), uint32(it.rowIdx)})
```

Comment:
Because it may read row from disk, should we cache the current row in memory?
So that we can avoid the extra disk read or deserialization for the multiple calls of `Current` in merge_join.go.

I've also noticed that `rowInDisk` has its own `bufReader`, we may do a benchmark to choose whether the cache of current row is necessary.

Reply:
Orz